### PR TITLE
56278 - download file issue

### PR DESCRIPTION
**Pull request by sumod13943**
Download logic changed

### CODE REVIEW

**Review by harshal11869 — APPROVED** [review]
LGTM.

**Review by ajasonca — CHANGES_REQUESTED** [review]
Just a minor simplification

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment/Program.cs [hunk 1]
```diff
@@ -177,7 +180,16 @@ private static void ConfigureServices(IServiceCollection serviceCollection, ICon
             serviceCollection.AddScoped<IZipHelper, ZipHelper>();
             serviceCollection.AddScoped<IFileUtility, FileUtility>();
 
-            serviceCollection.AddHttpClient();
+            serviceCollection.AddHttpClient("DownloadClient",
```
[the remark this answers is not in the document]
Comment (sumod13943) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment/Program.cs:183]: Changed

**Review by ajasonca — APPROVED** [review]
Looks good

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Helpers/FssApiClientTests.cs [hunk 2]
```diff
@@ -188,7 +188,7 @@ public void DoesDownloadFile_Returns_OK()
 
             _fakeFssApiClient = new FssApiClient(_fakeHttpClientFactory);
 
-            Task<HttpResponseMessage>? result = _fakeFssApiClient.DownloadFile("http://test.com", _authToken);
+            Task<HttpResponseMessage>? result = _fakeFssApiClient.DownloadFile("http://test.com", _authToken, "bytes=1-1024");
```
Comment (ray-sj) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Helpers/FssApiClientTests.cs:191]: please can we use fluent assertion

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Services/FssServiceTests.cs [hunk 3]
```diff
@@ -222,14 +221,13 @@ public async Task DoesDownloadFile_Returns_DownloadPath_If_ValidRequest()
                     },
                 });
 
-            Stream? result = await _fssService.DownloadFile("M01X02", "/batch/621e8d6f-9950-4ba6-bfb4-92415369aaee/files/M01X02.zip");
-
-            Assert.That(result, Is.Not.Null);
+            bool result = await _fssService.DownloadFile("M01X02", "/batch/621e8d6f-9950-4ba6-bfb4-92415369aaee/files/M01X02.zip", 10000, @"D:\POS");
```
Comment (ray-sj) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Services/FssServiceTests.cs:224]: please can we use fluent assertion

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Services/FssServiceTests.cs [hunk 4]
```diff
@@ -222,14 +221,13 @@ public async Task DoesDownloadFile_Returns_DownloadPath_If_ValidRequest()
...
 
+            Assert.That(result, Is.True);
```
Comment (ray-sj) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Services/FssServiceTests.cs:226]: why is there 2 asserts in a unit test, these need to focus on assertion of just a single point

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Services/FssServiceTests.cs [hunk 5]
```diff
@@ -249,7 +247,7 @@ public void DoesDownloadFile_Throws_Exception_If_InValidRequest()
...
                 });
 
-            Assert.ThrowsAsync<FulfilmentException>(() => _fssService.DownloadFile("M01X02", "/batch/621e8d6f-9950-4ba6-bfb4-92415369aaee/files/M01X02.zip"));
+            Assert.ThrowsAsync<FulfilmentException>(() => _fssService.DownloadFile("M01X02", "/batch/621e8d6f-9950-4ba6-bfb4-92415369aaee/files/M01X02.zip", 10000, @"D:\"));
```
Comment (ray-sj) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Services/FssServiceTests.cs:250]: multiple asserts in a single test

**Review by ray-sj — CHANGES_REQUESTED** [review]
There is no automated functional tests???

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Helpers/FssApiClientTests.cs (hunk 2 above)
Comment (sumod13943) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Helpers/FssApiClientTests.cs:191]: We have utilised Assert of Nunit throughout unit testing. From the subsequent project on, fluid assertion will be implemented.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Services/FssServiceTests.cs (hunk 4 above)
Comment (ajasonca) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Services/FssServiceTests.cs:226]: As long as the asserts are in a block of asserts this we think is acceptable. Therefore it must always be:

- Arrange
- Action
- Assert

So never do an action followed by asserts AND then do another action (e.g. see test Does_GetFileMD5_Returns_FileDetails_With_Hash which breaks this rule)

**Review by ray-sj — APPROVED** [review]
as discussed with the team and also walking through the unit test goals with Jason i'm happy that this is now covered.

**Pull request merged**